---
title: "Analysing DNA methylation data with Bioconductor"
author: "Peter Hickey ([@PeteHaitch](https://github.com/PeteHaitch/))"
date: "9 September 2015"
output: 
  ioslides_presentation:
    widescreen: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, comment = "#>")
library(biocViews)
library(BiocInstaller)
```

# Introduction

## Basic biology {.smaller}

<center><img src="images/5mC.png" height="480px" /></center>

Modified "<a href="https://commons.wikimedia.org/wiki/File:Cytosine_becomes_thymine.png#/media/File:Cytosine_becomes_thymine.png">Cytosine becomes thymine</a>" by <a href="//commons.wikimedia.org/wiki/User:CFCF" title="User:CFCF">CFCF</a> - <span class="int-own-work" lang="en">Own work</span>. Licensed under <a href="http://creativecommons.org/licenses/by-sa/3.0" title="Creative Commons Attribution-Share Alike 3.0">CC BY-SA 3.0</a> via <a href="//commons.wikimedia.org/wiki/">Wikimedia Commons</a>.

## Basic biology {.smaller}

<center><img src="images/Agouti-Mouse-1024x802.jpg" height="480px" /></center>

[http://episona.com/wp-content/uploads/2014/01/Agouti-Mouse-1024x802.jpg
](http://episona.com/wp-content/uploads/2014/01/Agouti-Mouse-1024x802.jpg
)

## Basic biology {.smaller}

<center><img src="images/methylation_vs_expression.png" height="480px" /></center>

[Figure S7.1C, Integrated genomic analyses of ovarian carcinoma, TCGA, doi:10.1038/nature10166](www.nature.com/articles/nature10166)

<div class = "notes">
Figure S7.1C. Scatterplots showing AMT gene expression versus promoter methylation. The color and size of the dots represent tissue type (red/large – fallopian tube samples, n=8; black/small – ovarian tumors, n=489). 
</div>

# Assays

## Assays {.smaller}

Focusing on modern high-throughput assays

### Microarrays

- Illumina 27k
- Illumina 450k

### Sequencing

- Bisulfite-treatment of DNA
    - Whole-genome bisulfite-sequencing
    - (Enhanced/Extended) Reduced representation
- Other
    - **M**ethylated DNA **i**mmuno**p**recipitation + sequencing (MeDIP-seq/mDIP-seq)
    - **M**ethylation-sensitive **r**estriction **e**nzyme + sequencing (MRE-seq)
    - **M**ethyl **b**inding **d**omain protein-enrichment + sequencing (MBD-seq)

## The analysis pipeline

__TODO: Make this slide into a picture (a line with "most specific" at one end and "least specific" at the other)__

Most assay-specific:

> - Getting data into R
> - Pre-processing
> - Analysis

Somewhere in between:

> - Batch effects

Less assay-specific:

> - Visualisation
> - Data integration

## How can Bioconductor help?

```{r methylation-bioc-pkgs, warning = FALSE}
data("biocViewsVocab")
# Get all biocViews
bioc_views <- lapply(X = BiocInstaller::biocinstallRepos()[1:4],
                     FUN = biocViews::getBiocViews,
                     vocab = biocViewsVocab,
                     defaultView = "NoViewProvided")
dna_methylation <- lapply(X = bioc_views,
                          FUN = function(bcv) bcv$DNAMethylation@packageList)
```

Bioconductor `r BiocInstaller::biocVersion()` packages (based on DNAMethylation [_BiocViews_](http://www.bioconductor.org/packages/release/BiocViews.html)):

> - `r length(dna_methylation[["BioCsoft"]])` software packages
> - `r length(dna_methylation[["BioCann"]])` annotation packages

## Manual package curation

__TODO: Working notes, not an actual slide__

Software: 

- Misc.
  - BEclear
  - bumphunter
  - COHCAP
  - coMET
  - MassArray
  - MEDIPS
  - MEDME
  - methVisual
  - methyAnalysis
  - methylMnM
  - Repitools
  - RnBeads
  
# Microarrays

## Description of assays

[Illumina 27k](https://en.wikipedia.org/wiki/Illumina_Methylation_Assay)

- Infinium HumanMethylation27 BeadChip
- Deprecated? No longer advertised on Illumina's website
- Much of TCGA data uses this platform
- Infinium I probes also on 450k

[Illumina 450k](http://www.biostat.jhsph.edu/~lcollado/misc/genomics/GWG_JPF.pdf)

- Infinium HumanMethylation450 BeadChip

## Key packages

- Microarrays
  - charm
  - conumee
  - CopyNumber450k
  - DMRcate
  - DMRforPairs
  - ENmix
  - lumi
  - MethylAid
  - MethylMix
  - methylumi
  - minfi
  - missMethyl
  - shinyMethyl
  - skewr
  - wateRmelon
  - __ANNOTATION__

## Analysis pipeline

Schematic of data generation + analysis pipeline

- Data format and ingest
- Pre-processing
  - Normalisation
  - CpG-SNPs
  - CNVs
- Downstream analyses
  - $\beta$-values vs. $\mathcal{M}$-values
  - Smoothing
  - Differential methylation: DMLs/DMPs, DMRs, kernel based tests
  - Differential variability


# Sequencing

## Description of assays

Bisulfite-sequencing-based assays

- WGBS
- [e]RRBS

Enrichment-based assays

- MeDIP-seq
- MREseq
- Not going to talk about much

## Key packages

- Sequencing
  - Bisulfite-sequencing
  - BiSeq
  - bsseq
  - DMRcaller
  - DSS
  - M3D
  - methylPipe
  - MethylSeekR
  - MPFE

## Analysis pipeline

- Non-R stuff: pre-processing, mapping, and post-processing of reads
  - List popular alignment software
  - M-bias
- Data input
- Pre-processing
  - CpG-SNPs
  - CNVs
  - Normalisation
- Downstream analyses
  - Smoothing
  - Beta-binomial model
  - Differential methylation: DMLs/DMPs, DMRs, kernel based tests.
  - Differential variability
  - Methylation patterns

# Visualisation

## Yup

- epiviz
- lollipop plots

# Data integration

## Yup 

- AnnotationHub, biomaRt, etc.

# Batch effects and technical biases

## Yup

- Cell composition bias

# Miscellaneous

## Yup

- Single-cell assays

# Links

## Yup

- Workflows
- Package vignettes
- Slides

## TODOs

- [ ] Fix up github link on title slide
- [ ] Include affiliation on title slide?
- [ ] Note that this is an opinionated talk
